node_ids: introduce GetNodeID

mhofstetter opened the pull request:
This PR introduces the possibility to retrieve the node id for a given node IP without having to use `AllocateNodeID` which comes with the drawback of actually allocating a new node id if one doesn't exist yet for the given IP.

Example: Using `ipcache.AllocateNodeID` to lookup the node ID for a node IP during auth gc initialisation results in unintended node ID allocations if the nodeids doesn't exist yet for the cilium nodes.

By using the new method `GetNodeID` we remove this unintended side-effect.

Therefore, the temporarily disabled auth map GC functionality has been re-enabled!

Related to: https://github.com/cilium/cilium/pull/26073 & https://github.com/cilium/cilium/issues/25964

mhofstetter committed:
node_ids: introduce GetNodeID

This commit introduces the possibility to retrieve the node id for a
given node IP without having to use `AllocateNodeID` which comes with
the drawback of actually allocating a new node id if it doesn't exist
yet.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
auth: replace ipcache.AllocateNodeID with ipcache.GetNodeID

Using `ipcache.ALlocateNodeID` to lookup the node id for a node IP
during auth gc initialisation results in unintended node id allocations
if the nodeids aren't yet created for the cilium nodes.

By using the new method `GetNodeID` we remove this unwanted side-effect.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
node_ids: re-use fake nodeidhandler in wireguard test

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
auth: re-enable node-based auth map gc

This commit re-enables the node-based auth map garbage collection which
have been temporarily disabled with #26073.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter commented on pkg/datapath/linux/node_ids.go:
```diff
@@ -77,8 +72,28 @@ func (n *linuxNodeHandler) GetNodeIP(nodeID uint16) string {
 	return n.nodeIPsByIDs[nodeID]
 }
 
+func (n *linuxNodeHandler) GetNodeID(nodeIP net.IP) (uint16, bool) {
```
do we want to go with an `error` instead of the `bool` in the "public" facing API? my thoughts were that a `bool` fulfills the purpose.

meyskens replied:
I agree a bool will be fine as it mimics the map "interface" just fine

mhofstetter commented on pkg/wireguard/agent/agent_test.go:
```diff
@@ -82,7 +83,7 @@ func containsIP(allowedIPs []net.IPNet, ipnet *net.IPNet) bool {
 func newTestAgent(ctx context.Context) (*Agent, *ipcache.IPCache) {
 	ipCache := ipcache.NewIPCache(&ipcache.Configuration{
 		Context:       ctx,
-		NodeIDHandler: &mockNodeHandler{},
```
idea: re-using the "duplicate" from ipcache kind of reduces the blast radius (e.g. wireguard) when changing the nodeidhandler interface

meyskens reviewed (approved):
LGTM for me

mhofstetter committed:
node: use RWMutex for linuxNodeHandler GetNodeIP & GetNodeID

Until now, a `lock.Mutex` was securing all fields of the
linuxnodehandler.

With the introduction ready-only methods GetNodeIP & GetNodeID, it
became useful to replace it with an RWMutex and only lock it for read in
these two functions.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
node_ids: return 0 for IPv6 in getNodeIDForIP too

Currently, when retrieving the node id for an IP, the local node id 0 is
only returned if the given IP matches the nodes IPv4 - but not IPv6.

This commit adds support for IPv6.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

jrajahalme reviewed (approved):
LGTM

jibi commented on pkg/datapath/linux/node_ids.go:
```diff
@@ -25,24 +25,19 @@ const (
+// Don't allocate a node ID for the local node.
 func (n *linuxNodeHandler) AllocateNodeID(nodeIP net.IP) uint16 {
 	if len(nodeIP) == 0 || nodeIP.IsUnspecified() {
-		// This should never happen. If it ever does, we may have an unexpected
```
should this comment stay here?

mhofstetter replied:
@jibi good catch - yes it should stay here - thanks. seems to be reordered when moving this check (&warning) back to `AllocateNodeID` because it doesn't make sense to have this in `getNodeIDForIP`

mhofstetter replied:
moved the comments back to its original place

jibi commented on pkg/datapath/linux/node_ids.go:
```diff
@@ -77,8 +72,28 @@ func (n *linuxNodeHandler) GetNodeIP(nodeID uint16) string {
+func (n *linuxNodeHandler) getNodeIDForIP(nodeIP net.IP) (uint16, bool) {
+	localNode := node.GetIPv4()
+	if localNode.Equal(nodeIP) {
+		return 0, true
```
why are we returning true if we are not suppose to allocate a node ID for the local node?

mhofstetter replied:
@jibi `0` is still the node id for the local node (and therefore has its meaning for users of this (e.g. whether not to allocate in `AllocateNodeID` or not just getting the info that its the local node itself via `GetNodeID` ).
